### PR TITLE
hawk: Add counter feature using PERCPU_ARRAY map

### PR DESCRIPTION
**Pull request by Poppy22**
- added PERCPU_ARRAY map to kernel space to monitor the number of processes executed on each process
- added a loop that prints the total count in user space once every second
- updated readme and added comments

### CODE REVIEW

**Comment by Poppy22** [comment]
Reviewers: @FlorentRevest @sinkap

File: .github/README.md [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 ## 1. Overview
 
-In this short tutorial you will learn how to write and compile a simple **LSM BPF program** that performs **bpf_trace_printk**. The program will be loaded into the kernel and will print a string after the execution of each process.
+In this short tutorial you will learn how to write and compile a simple **LSM BPF program** that performs **bpf_trace_printk**. The program will be loaded into the kernel and will print a string after the execution of each process. Moreover, the program uses a BPF map of type PERCPU_ARRAY to keep track of the number of processes for each CPU, printing the total count each second.
```
Comment (FlorentRevest) [inline .github/README.md]: This makes it sound like the BPF program is printing the total count.
I would reword it slightly with s/printing the/allowing userspace to print/

File: .github/README.md [hunk 2]
```diff
@@ -84,13 +84,13 @@ Now compile exec.c:
 kcc exec.c
 ```
 
-Go to src/user and generate `exec.skel.h` with the following command:
+Now, still from src/kern,  generate `exec.skel.h` with the following command:
```
Comment (FlorentRevest) [inline .github/README.md:87]: A best-practice with git (or any source control) is to keep changes "atomic" (minimal). Ideally, each commit does one thing and only one. There are good reasons for this. One reason is that I might be looking at the project git log in search of a commit that could have introduced this change, and if I see "hawk: Add counter feature using PERCPU_ARRAY map" I will instantly assume that it did not change that line because that seems unrelated. Another reason is that someday we might discover that this PERCPU_ARRAY caused a massive bug and we want to revert the code asap. If each commit is atomic, we can easily use `git revert commitHash` and drop only the feature that caused the bug. But in that case, if we were to revert the PERCPU_ARRAY we would also loose the README refactoring.

One last tip to make this easier. You can use "git add -p README.md" to be prompted with an interactive tool that lets you pick what changes in the file go into the next commit. For example, if you have already a README.md with changes all around, you can use `git add -p` to select the refactoring once, then git commit and then later create a new commit with what remains.

File: src/kern/exec.c [hunk 3]
```diff
@@ -17,11 +17,42 @@
...
+         * val points to output_map[key][i], where i is the index of the current
+         * CPU; this index is already known by the BPF program, as it internally
+         * keeps track of the CPUs.
+         */
+        int *val = bpf_map_lookup_elem(&output_map, &key);
+        if (!val) {
+                return 0;
+        }
+
+        // Increment the previous value, as a new process was executed the
```
Comment (FlorentRevest) [inline src/kern/exec.c]: executed on* the current CPU :)

File: src/user/user.c [hunk 4]
```diff
@@ -1,6 +1,6 @@
 /*
  * Copyright 2020 Google LLC
- * Licensed under the Apache License, Version 2.0 (the "License");
+ * Licensed under the Apache License, Version 3.0 (the "License");
```
Comment (FlorentRevest) [inline src/user/user.c]: mh?
There are no Apache 3 licenses https://www.apache.org/licenses/
And if there was, you should update all references to it (for example 4 lines below you link to the LICENSE-2.0)

File: src/user/user.c [hunk 5]
```diff
@@ -13,12 +13,19 @@
...
+#include <bpf/libbpf.h>
 #include "exec.skel.h"
 #include <unistd.h>
+#include <bpf/bpf.h>
 
 int main(int ac, char **argv)
 {
+        int cpus = libbpf_num_possible_cpus(); // the number of CPUs on your machine
+        unsigned long long values[cpus]; // number of executed processes on each CPU
+        int  map_fd;
```
Comment (FlorentRevest) [inline src/user/user.c]: extra space between int and map_fd

File: src/user/user.c [hunk 6]
```diff
@@ -13,12 +13,19 @@
...
-#include <linux/bpf.h>
+#include <bpf/libbpf.h>
 #include "exec.skel.h"
 #include <unistd.h>
+#include <bpf/bpf.h>
 
 int main(int ac, char **argv)
 {
+        int cpus = libbpf_num_possible_cpus(); // the number of CPUs on your machine
+        unsigned long long values[cpus]; // number of executed processes on each CPU
```
Comment (FlorentRevest) [inline src/user/user.c]: Usually, if a variable name is well chosen, it does not need a comment. (which is helpful because I might start reading the code later in the file and stumble upon this "value" and I would be really unsure what it is, that would force me to go back and read the comment)
This is just a proposal but executions_nb_on_cpu does not need a comment and when you use it `executions_nb_on_cpu[3]` it reads more easily than `value[3]`

Comment (FlorentRevest) [reply src/user/user.c]: Similar things can be said for other variables, maybe cpus_nb makes more sense than cpus, because it avoids one comment too.

File: src/user/user.c [hunk 7]
```diff
@@ -35,8 +42,31 @@ int main(int ac, char **argv)
                 return 0;
         }
 
-        sleep(20);
+        // get the fd for the map created in the BPF program
+        map_fd = bpf_map__fd(skel->maps.output_map);
+
+        // count the total number of processes executed on the CPUs each second
+        for (int i = 0; i < 1000; i++) {
```
Comment (FlorentRevest) [inline src/user/user.c]: This feels a bit arbitrary. Maybe I love looking at the output of this program and would be disappointed that after 1000 seconds it stops printing numbers. What about `while (1)` ?

File: src/user/user.c [hunk 8]
```diff
@@ -35,8 +42,31 @@ int main(int ac, char **argv)
...
+                /*
+                 * Search for the element with key = 0 in the map.
+                 * values will point to output_map[key]
+                 */
+                bpf_map_lookup_elem(map_fd, &key, values);
+
+                // reset the counter to avoid counting duplicates
+                counter = 0;
+
+                // loop over the values, add them up and reset the value of each CPU
```
Comment (FlorentRevest) [inline src/user/user.c]: This is no longer true, you do not reset the value of each CPU

File: src/kern/exec.c [hunk 9]
```diff
@@ -17,11 +17,42 @@
...
 #include <bpf/bpf_tracing.h>
 
+/*
+ * Define a map structure to communicate with the userspace program.
+ * The type of map is BPF_MAP_TYPE_PERCPU_ARRAY;
+ * it contains one element, found for key = 0
+ * and the value is an array with number_of_CPUs elements.
+ */
+struct bpf_map_def SEC("maps") output_map = {
+        .type = BPF_MAP_TYPE_PERCPU_ARRAY,
```
Comment (FlorentRevest) [inline src/kern/exec.c]: I see that you are using 8 spaces for indentation, but the Linux kernel style actually requires "tabulation" characters. They recommend that you configure your editor to show tabs as 8 spaces because there are times when you want to align something using spaces (eg if your function arguments span two lines, the second line should be aligned with the first one using tabs and then spaces) but your code should actually not contain spaces.

I went back to look at the style guide and I reckon it is really stupidly worded: "Outside of comments, documentation and except in Kconfig, spaces are never used for indentation, and the above example is deliberately broken." so basically they give you an example but you shouldn't follow it, aerrm.

If you look at a random function, say https://github.com/torvalds/linux/blob/master/kernel/torture.c#L204 you can see that indentation is done with one character (tab)

Anyway, let's fix this separately, this is one of those things that deserve a separate commit (or PR) to keep each commit "atomic" as I explained above.

File: src/kern/exec.c [hunk 10]
```diff
@@ -17,11 +17,42 @@
...
+        .key_size = sizeof(int),
+        .value_size = sizeof(int),
+        .max_entries = 1,
+};
+
 SEC("lsm/bprm_committed_creds")
 int BPF_PROG(test_void_hook, struct linux_binprm *bprm)
 {
         char str[] = "hello\n";
+        int key = 0; // the only key in the map, since there is just one pair key-value
```
Comment (FlorentRevest) [inline src/kern/exec.c]: one key-value pair*?

**Review by FlorentRevest — CHANGES_REQUESTED** [review]

File: src/user/user.c (hunk 6 above)
Comment (FlorentRevest) [reply src/user/user.c]: This is actually part of the linux style guide https://www.kernel.org/doc/html/v4.10/process/coding-style.html#commenting over-commenting code does not help. Also a good comment describes what the function does, not how.

One good reason for this is that the "what" shouldn't change over time, but the "how" might change and most likely your comment will be left outdated (cf your "loop over the values, add them up and reset the value of each CPU" which was already outdated before you even submitted the PR :) )

Another good reason is that the reader is expected to know C. I expect anyone who glances at:
```
for (int i = 0; i < cpus; i++) {
        counter += values[i];
}
```
To figure out that this "loop over the values and add them up" in a split of a second. What might help the reader though is what are those "values", and why are they being added up? Maybe something like "Accumulate execution number on each CPU" is already more helpful. And maybe someday the implementation will change to something else but the comment will stay relevant.

File: src/user/user.c [hunk 11]
```diff
@@ -35,8 +42,30 @@ int main(int ac, char **argv)
                 return 0;
         }
 
-        sleep(20);
+        // get the fd for the map created in the BPF program
+        map_fd = bpf_map__fd(skel->maps.output_map);
+
+        // count the total number of processes executed on the CPUs each second
+        while (1) {
+
```
Comment (FlorentRevest) [inline src/user/user.c:55]: Empty lines should convey meaning by separating logical blocks, in this case it does not have any value so I would just delete that line. (I'm yet to see an empty line at the beginning of a block that has meaning)

**Review by FlorentRevest — APPROVED** [review]
